gcp_variant_transforms/libs/variant_partition.py

```diff
@@ -34,26 +43,216 @@
+      raise ValueError('Index of a region cannot be negative {}'.format(index))
+    self._index = index
+
+  def get_index(self, pos=0):
+    if self._interval_tree is None and self._index == -1:
+      raise ValueError('This instance has not been assigned any index.')
+    if self._interval_tree is None:
+      return self._index
+    else:
+      matched_intervals = self._interval_tree.search(pos)
+      if len(matched_intervals) > 1:
+        raise ValueError('Search for pos {} returned more than 1 result'.
+                         format(pos))
+      if len(matched_intervals) == 1:
+        return next(iter(matched_intervals)).data
+      else:
+        return -1
+
+  @property
+  def is_full_chromosome(self):
+    return self._index != -1
+
+class VariantPartition(object):
+  """Partition variants based on their reference_name and/or position.
+
+  This class has 2 operating modes:
+    1) No config file is given (config_file_path == ''):
+        Automatically partition variants based on their reference_name.
+    2) A config file is given:
+        partitions variants based on input config file.
+  """
+
+  def __init__(self, config_file_path=''):
+    if not config_file_path:
+      if _DEFAULT_NUM_PARTITIONS <= _RESERVED_AUTO_PARTITIONS:
+        raise ValueError(
+            '_DEFAULT_NUM_PARTITIONS must be > _RESERVED_AUTO_PARTITIONS')
+      self._config_file_path_given = False
+      self._reference_name_to_partition_map = {}
+    else:
+      self._config_file_path_given = True
+      self._parse_config(config_file_path)
+
+  def _parse_config(self, config_file_path):
+    """Parses the given partition config file.
+
+    Args:
+      config_file_path: name of the input partition_config file.
+    Raises:
+      A ValueError if any of the expected config formats are violated.
+    """
+    def _parse_position(pos_str):
+      return int(pos_str.replace(',', ''))
+
+    with FileSystems.open(config_file_path, 'r') as f:
+      try:
+        partition_configs = yaml.load(f)
+      except yaml.YAMLError as e:
+        raise ValueError('Invalid yaml file: %s' % str(e))
+
+    if len(partition_configs) > _MAX_NUM_PARTITIONS:
       raise ValueError(
-          '_DEFAULT_NUM_PARTITIONS must be > _RESERVED_AUTO_PARTITIONS')
-    self._reference_name_to_partition_map = {}
+          'There can be at most {} partitions but given config file contains {}'
+          .format(_MAX_NUM_PARTITIONS, len(partition_configs)))
+    if not partition_configs:
+      raise ValueError('There must be at least one partition in config file.')
+
+    self._num_partitions = len(partition_configs)
+
+    self._by_ref_name = defaultdict(RegionIndexer)
+
+    self._suffixes = []
+    table_suffix_duplicate = {}
+
+    # Default partition will contain all remaining readings that do not match to
+    # any other partition. -1 means it does not exist.
+    self._default_partition_index = -1
+    partition_index = 0
+    for partition_config in partition_configs:
```

Comment:
I think this logic can be simplified a lot. Let's discuss later on how we can do it.

Reply:
I tried to simplify the logic, let me know what you think.